src/main/java/org/springframework/integration/kafka/listener/MetadataStoreOffsetManager.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+
+package org.springframework.integration.kafka.listener;
+
+import static com.gs.collections.impl.utility.Iterate.groupBy;
+
+import java.io.Closeable;
+import java.io.Flushable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import com.gs.collections.api.RichIterable;
+import com.gs.collections.api.block.function.Function;
+import com.gs.collections.api.block.procedure.Procedure2;
+import com.gs.collections.api.map.MutableMap;
+import com.gs.collections.api.multimap.MutableMultimap;
+import com.gs.collections.impl.map.mutable.ConcurrentHashMap;
+import kafka.api.OffsetRequest;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.integration.kafka.core.BrokerAddress;
+import org.springframework.integration.kafka.core.ConnectionFactory;
+import org.springframework.integration.kafka.core.KafkaConsumerDefaults;
+import org.springframework.integration.kafka.core.Partition;
+import org.springframework.integration.kafka.core.Result;
+import org.springframework.integration.metadata.MetadataStore;
+import org.springframework.integration.metadata.SimpleMetadataStore;
+import org.springframework.util.CollectionUtils;
+
+/**
+ * An {@link OffsetManager} that persists offsets into a {@link MetadataStore}.
+ * This offset manager maintains a transient internal state, which is the
+ *
+ * @author Marius Bogoevici
+ */
+public class MetadataStoreOffsetManager implements OffsetManager {
+
+	private final static Log log = LogFactory.getLog(MetadataStoreOffsetManager.class);
+
+	private final GetLeaderFunction getLeaderFunction = new GetLeaderFunction();
+
+	private String consumerId = KafkaConsumerDefaults.GROUP_ID;
+
+	private MetadataStore metadataStore = new SimpleMetadataStore();
+
+	private ConnectionFactory connectionFactory;
+
+	private long referenceTimestamp = KafkaConsumerDefaults.DEFAULT_OFFSET_RESET;
+
+	private MutableMap<Partition, Long> offsets = new ConcurrentHashMap<Partition, Long>();
```

Comment:
Clusters are supported if the `MetadataStore` supports them. The internal state is intentionally transient, and we never intend to support multiple concurrent consumers on the same topic/partition (unless that is a result of fanout in the `MessageListener`). In other words, updates made via `updateOffset` are always propagated to the persistent storage.  

Detailed explanations have been provided in the following comments: https://github.com/spring-projects/spring-integration-kafka/pull/12#discussion_r22799956, https://github.com/spring-projects/spring-integration-kafka/pull/12#discussion_r22800652 and https://github.com/spring-projects/spring-integration-kafka/pull/12#discussion_r22808961